[Docs] Explicitly document libclang ABI and API stability

AaronBallman opened the pull request:
Our current docs leave a lot of latitude ("relatively stable") without explaining what the goals are for this stability. This patch adds some basic documentation explaining that there are some changes which can impact ABI and API stability that we reserve the right to make, lists some scenarios we explicitly do not support, but otherwise tries to assure the reader that the APIs and ABI are stable.

AaronBallman committed:
[Docs] Explicitly document libclang ABI and API stability

Our current docs leave a lot of latitude ("relatively stable") without
explaining what the goals are for this stability. This patch adds some
basic documentation explaining that there are some changes which can
impact ABI and API stability that we reserve the right to make, lists
some scenarios we explicitly do not support, but otherwise tries to
assure the reader that the APIs and ABI are stable.

AaronBallman commented:
I've added a handful of folks who might be reasonable reviewers for this, but if you know of anyone else who may have an opinion, feel free to rope them in.

These changes came about because of https://github.com/llvm/llvm-project/pull/134551 where someone was trying to add a better string API that handles embedded nulls but we ran into "ABI stability" when working through the implementation.

The intent here is not to shift away from ABI stability for the library but to remind folks that we still need enough wiggle room to maintain and evolve it, and that underlying details of Clang can sometimes still mean breaking changes between libclang versions.

DeinAlptraum reviewed (approved):
This looks reasonable to me, but I'm coming from the Python bindings anyway and can't comment on the details for the C/C++ side.

Endilll commented on clang/docs/LibClang.rst:
```diff
@@ -358,3 +358,46 @@ Complete example code
+The library should be ABI and API stable over time, but ABI- and API-breaking
+changes can happen in the following situations:
+
+* Adding new enumerator to an enumeration (can be ABI-breaking in C++).
+* Removing an explicitly deprecated API after a suitably long deprecation
+  period.
+* Uses of implementation details, such names or comments that say something is
+  "private", "reserved", "internal", etc.
```
> Is this list meant to be exhaustive? If not, perhaps the wording should be changed.

I think "etc." at the end conveys that it's not exhaustive

> I'm coming only from the Python-bindings perspective, but we did have a number of "potentially breaking changes" 

Yeah, bindings for specific languages need additional docs, because they are not as conservative as C API.

AaronBallman committed:
Update some typos found during code review

AaronBallman committed:
Minor grammar fix

AaronBallman committed:
Make it more clear the list is not exhaustive, talk about bug fixes

AaronBallman committed:
Slight tweak

I had an offline discussion with Endilll and we came up with some small
adjustments.
